Add LDL decomposition

ajefweiss opened the pull request:
See issue #865, this decomposition is also mentioned in #762 but in the end only the UDU decomposition was implemented (PR #766). I need this for my own use case where I basically want a Cholesky decomposition of a semi-definite covariance matrix to generate random correlated numbers. Currently I was applying the Cholesky decomposition on a modified matrix with all zero rows/columns removed, which are then later re-added to the lower triangular matrix.

I used the code from #766 as a template, but changed the code to implement the LDL algorithm instead, everything else is basically the same just re-named. This is not the standard algorithm, but a variant that should also work with semi-definite matrices (i.e. with eigenvalues equal to zero), which I need. I haven't added any tests for this yet though. I do not know why the existing UDU decomposition also does not include this capability as its just an extra 1-2 lines of code.

I also added a `cholesky_l` function that reconstructs the lower triangular matrix from the Cholesky decomposition.

ajefweiss committed:
basic ldl implementation, based on udu

ajefweiss committed:
test for cholesky eq

alexhroom commented:
hi, looking forward to this being added - I was wondering whether it's necessary for it to insist on the scalar type having trait `RealField`? In the UDU factorisation PR it's set to `RealField` because nobody was sure whether the UDU decomposition made sense for complex Hermitian matrices, but LDL is pretty widely used for complex matrices.

I tested it out on a few complex Hermitian matrices and got the results I was expecting. Maybe the functionality could be extended to `ComplexField` and you add some tests for equality with Cholesky on positive-definite complex Hermitian examples?

ajefweiss committed:
hermitian ldl, better tests

ajefweiss commented:
I added functionality for Hermitian matrices, and rewrote the tests.

Currently, the algorithm checks for diagonal values that are exactly zero, to zero out the corresponding columns/rows. One could add a threshold value `eps`, similar to the way its done for the pseudo_inverse() function, so that values below `eps` are set to zero. Only problem here is that the diagonal values have no real relation to the eigenvalues, so the interpretation is not clear.

alexhroom commented:
@ajefweiss cool! for clarity i'm not a maintainer/reviewer of `nalgebra` i'm just interested in having this in the package :)

geo-ant commented:
Hi @ajefweiss, I'd be willing to review this but I need your help :slightly_smiling_face: I skimmed the paper linked in https://github.com/dimforge/nalgebra/issues/762, but I couldn't find an algorithm for the LDL decomposition. I see there's a screenshot in the issue for UDU decomposition. Could you point me to the reference that tells me how you implemented the LDL decomposition, so I can read up on it?

ajefweiss commented:
Hi @geo-ant!

Thank you for taking a look at this. The actual reference I used is just the wikipedia entry which has imo a nicer description of the algorithm (https://en.wikipedia.org/wiki/Cholesky_decomposition#LDL_decomposition_2) which cites "Fundamentals of Matrix Computations" from Watkins (PDF versions are easily obtainable online). The relevant pages here are 82-84 although the algorithm here is just explained as a variant of the LU decomposition without being explicitly written out.

The only problem is that I have not been able to find an explicit reference for an algorithm that allows for semi-positive definite matrices. Both the reference you linked, and mine, require positive-definite matrices in their algorithms due to the division by v_i / d_i. Thus when one of the diagonal entries is 0 the algorithm fails. From my understanding, the A=LL/LDL factorizations are unique if, and only if, the matrix A is positive-definite. The theorems from both references state that the LDL factorization does exist only under the requirement that A is symmetric (and not necessarily positive definite). 

I think the factorization is no longer unique because you can fill out the relevant column in the L matrix with any values as they will be subsequently zero'd out in the multiplication. In my code this column is filled with zeros.

PS: I might try to think a better name for the function "cholesky_ldl" as I actually don't think that name is appropriate and might be a bit confusing.

geo-ant commented on tests/linalg/ldl.rs:
```diff
@@ -0,0 +1,108 @@
+                proptest! {
+                    #[test]
+                    fn ldl(m in dmatrix_($scalar)) {
+                        let m = &m * m.adjoint();
```
If I understood correctly what you said in the comment, then this implementation of the LDL^T decomposition requires the matrix to be symmetric positive *definite* (not *semi* definite), right? However, using `M M^T` will only give us pos. semidefinite, so we should expect some spurious failures here. I've actually run into the same problem in `nalgebra-lapack` and you can see my solution [here](https://github.com/dimforge/nalgebra/blob/main/nalgebra-lapack/tests/linalg/cholesky.rs). I've added a proptest function that gives a spd matrix in a bit of a hacky, but sound way: The idea is to use `M' = M M^T + alpha * Id`, where alpha is chosen suitably large to make we don't get into numerical trouble. You should be able to easily adapt the functions to return complex matrices.

alexhroom replied:
just wanted to weigh in - i'm a little confused actually... the 'point' of LDL is that it's a generalisation of the Cholesky algorithm to **all** symmetric/Hermitian matrices, with no requirements on positive definiteness/semi-definiteness. If it only works for definite/semidefinite matrices then there's no benefit to using LDL over Cholesky, which is a simpler decomposition _and_ a more efficient algorithm.

geo-ant replied:
Hey @awinick, thanks for weighing in! Do you have a source for the symmetric indefinite LDL decomposition that you mentioned?

awinick replied:
Hey @geo-ant you might be interested in the PR I just opened.

geo-ant commented on src/linalg/ldl.rs:
```diff
@@ -0,0 +1,102 @@
+                    d_j -= l[(j, k)].clone() * l[(j, k)].clone().conjugate() * d[k].clone();
+                }
+            }
+
```
I've thought about the algorithm a bit and I'm wondering if you'd mind implementing the algorithm in https://github.com/dimforge/nalgebra/issues/762#issuecomment-694735837. The reason is, that algorithm should be a  more efficient in terms of memory usage (and that might make it more cache friendly, but I'm not sure) than your implementation. The reason is that the linked algorithm works in place, while your algorithm works out of place. Your interface wouldn't even have to change because you're taking the OMatrix by value already. Your original matrix A would then contain the LDL decomposition where L is in the stricly lower triangular part and d is on the diagonal.

The only problem I see then, is that exposing access to L and D would require new allocations. Is that why you implemented the algorithm out of place in the first place?

What I mean is this: is the primary use of LDL^T decomposition to get explicit access to L and D, or would it also be used to solve linear systems. In that case we could expose a solve method that uses the triangular solvers internally and deals with D through scaling.

ajefweiss replied:
Code has ben re-factored to make the calculations in-place, I'm still using the same algorithm just not allocating new matrices.

geo-ant reviewed (commented):
thanks for the great work, see some comments in the review. 

You said that this algorithm _should_ also exist for positive **semi** definite matrices, but your implementation only covers positive definite ones. As long as we document this, I don't see a problem here.

alexhroom commented on src/linalg/ldl.rs:
```diff
@@ -0,0 +1,102 @@
+use crate::dimension::Dim;
+use simba::scalar::ComplexField;
+
+/// LDL factorization.
```
just want to add to make sure we clarify that `D` is **strictly diagonal** - people also use LDL to refer to a different algorithm which decomposes a Hermitian matrix into a lower unit-triangular matrix L and a **block-diagonal** matrix `D` composed of 1x1 and 2x2 blocks. 

This is the algorithm implemented in the [*HETRF family of algorithms in LAPACK](https://netlib.org/lapack/explore-html/d8/d0e/group__hetrf_ga9e7136edefc49152fe75a2c2bc410bce.html), which are [wrapped by SciPy's `ldl` function in Python](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.ldl.html) as well as [MATLAB's `ldl` function](https://www.mathworks.com/help/matlab/ref/ldl.html), so users may be expecting the block-diagonal algorithm and be confused. (the idea of the block-diagonal algorithm is that there are bounds on the size of the elements of `D` which do not exist for the strictly diagonal algorithm)

 The Rust package `faer` has both versions - they call the strictly diagonal one LDLT and the block-diagonal one LBLT.

> Thinking about this some more, do you think this decomposition should be called LDL or LDLT? I'm unsure...

for some comparisons summarising info above, both [Eigen (C++)](https://libeigen.gitlab.io/eigen/docs-nightly/classEigen_1_1LDLT.html) and [`faer` call it LDLT](https://docs.rs/faer/latest/faer/mat/type.Mat.html#method.ldlt), [SciPy calls it LDL](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.ldl.html), as does [MATLAB](https://www.mathworks.com/help/matlab/ref/ldl.html).

ajefweiss replied:
Thanks for the comments. 

Correction:  My implementation of the algorithm covers semi-positive (and semi-negative) definite matrices. The referenced algorithms in the textbooks DO NOT.

Regarding the naming convection LDL/LDLT. As @alexhroom pointed out, both conventions are used. Given the fact that we already have the UDU factorization implemented as UDU (and not UDU^T), I think it would make sense to keep it as LDL.

ajefweiss committed:
update docs, renamed cholesky_l to lsqrt and added extra checks

ajefweiss committed:
in place ldl factorization

ajefweiss commented:
I am planning to take a look at this again this weekend and clean up the remaining documentation. I believe that having both implementations should be useful as the strictly diagonal LDL version should be quicker if the assumed constraints hold (I plan on testing this).

There is only one thing I think that needs to be resolved regarding the naming convention. LAPACK has both LDL / UDU implemented incl. using the Bunch-Kaufman algorithm. But algebra nalready now has the strictly diagonal version for UDU already implemented. It might be confusing to call the BK implementation just "LDL" as it behaves differently than the existing UDU.

awinick commented:
@ajefweiss One thought on naming: since this is the strictly diagonal / no-pivot variant, would something like `Qdldl` make sense here? That seems to make the restricted contract clearer than plain `LDL`.

Also, would it make sense to detect breakdown explicitly instead of zeroing the column and continuing when a diagonal pivot is zero? The [[QDLDL implementation](https://github.com/osqp/qdldl/blob/master/src/qdldl.c)](https://github.com/osqp/qdldl/blob/master/src/qdldl.c) feels fairly close in spirit here and returns failure on zero pivots.

If semidefinite-ish cases are still important, maybe that behavior could be configurable via an API choice or pivot tolerance.

Curious what you think.

geo-ant commented:
Hey, @ajefweiss and @awinick, first of all sorry to both of you for not reviewing your implementations so far, life got in the way...

I'm happy you are already discussing the tradeoffs of having both implementations in nalgebra. In terms of naming, I think we could follow the suggestion of @alexhroom and do it like in `faer`, while also making sure that the documentation is very explicit about what the decomposition does and doesn't provide.

ajefweiss committed:
ldl docs and unchecked

ajefweiss commented:
I have added extra documentation pointing out the limitation and also referred to the BK algorithm. Until that's merged I don't think I have a good way of creating a proper link to those types but that's always something we can clean up later.

Renamed d/l functions to have the same names as in the upcoming LBL implementation.

LDL::new now has the "expected" behavior and returns None if there is a zero pivot. 

I also added a LDL::new_unchecked for the special case with zero pivots. This decomposition may or may not return a valid LDL decomposition depending on the structure of the matrix, always works for covariance matrices. I have not added this function in the decomposition.rs file

ajefweiss committed:
ldl determinant, solve, solve_mut

ajefweiss committed:
Merge branch 'main' into ldldecomp

ajefweiss committed:
fixed statement in the docs
